File: calendar/solar/generate_orbit_angles.py
import pandas as pd
from datetime import datetime, timedelta
from skyfield.api import load
from skyfield import almanac

# Planets to track
PLANETS = [
    'mercury',
    'venus',
    'earth',  # for reference
    'mars',
    'jupiter barycenter',
    'saturn barycenter',
    'uranus barycenter',
    'neptune barycenter',
]

def summer_solstice_first_monday(eph, ts, year):
    """Return Time of the first Monday after the June solstice."""
    t0 = ts.utc(year, 6, 20)
    t1 = ts.utc(year, 7, 1)
    t, y = almanac.find_discrete(t0, t1, almanac.seasons(eph))
    solstice = None
    for ti, yi in zip(t, y):
        if yi == almanac.SUMMER:
            solstice = ti.utc_datetime().date()
            break
    if solstice is None:
        raise ValueError('Could not determine summer solstice')

    d = datetime.combine(solstice, datetime.min.time())
    while d.weekday() != 0:  # 0 = Monday
        d += timedelta(days=1)
    return ts.utc(d.year, d.month, d.day)


def compute_orbits(year=2024):
    ts = load.timescale()
    eph = load('de421.bsp')

    start_time = summer_solstice_first_monday(eph, ts, year)

    sun = eph['sun']
    start_long = sun.at(start_time).ecliptic_position().longitude.degrees

    days = 364
    records = []
    for day in range(days):
        t = start_time + day
        row = {'day': day}
        for name in PLANETS:
            body = eph[name]
            lon = body.at(t).ecliptic_position().longitude.degrees
            rel_angle = (lon - start_long) % 360
            row[name] = rel_angle
        records.append(row)

    df = pd.DataFrame(records)
    return df

if __name__ == '__main__':
    df = compute_orbits()
    df.to_csv('orbit_angles.csv', index=False)
    print('Saved orbit_angles.csv')

File: calendar/solar/generate_orbit_angles.js
const Astronomy = require('astronomy-engine');
const fs = require('fs');

function summerSolsticeFirstMonday(year) {
  const seasons = Astronomy.Seasons(year);
  const solsticeDate = seasons.jun_solstice.date;
  const d = new Date(Date.UTC(solsticeDate.getUTCFullYear(), solsticeDate.getUTCMonth(), solsticeDate.getUTCDate()));
  while (d.getUTCDay() !== 1) { // Monday = 1
    d.setUTCDate(d.getUTCDate() + 1);
  }
  return new Astronomy.AstroTime(d);
}

function computeOrbits(year = 2024) {
  const start = summerSolsticeFirstMonday(year);
  const sunVec = Astronomy.GeoVector(Astronomy.Body.Sun, start, false);
  const sunLon = Astronomy.Ecliptic(sunVec).elon;

  const planets = [
    { body: Astronomy.Body.Mercury, name: 'mercury' },
    { body: Astronomy.Body.Venus, name: 'venus' },
    { body: Astronomy.Body.Earth, name: 'earth' },
    { body: Astronomy.Body.Mars, name: 'mars' },
    { body: Astronomy.Body.Jupiter, name: 'jupiter' },
    { body: Astronomy.Body.Saturn, name: 'saturn' },
    { body: Astronomy.Body.Uranus, name: 'uranus' },
    { body: Astronomy.Body.Neptune, name: 'neptune' },
  ];

  const records = [];
  const days = 364;
  for (let day = 0; day < days; day++) {
    const t = start.AddDays(day);
    const row = { day };
    for (const p of planets) {
      const vec = Astronomy.GeoVector(p.body, t, false);
      const lon = Astronomy.Ecliptic(vec).elon;
      const rel = (lon - sunLon + 360) % 360;
      row[p.name] = rel.toFixed(6);
    }
    records.push(row);
  }
  return records;
}

if (require.main === module) {
  const data = computeOrbits();
  const headers = Object.keys(data[0]);
  const lines = [headers.join(',')];
  for (const row of data) {
    lines.push(headers.map(h => row[h]).join(','));
  }
  fs.writeFileSync('orbit_angles.csv', lines.join('\n'));
  console.log('Saved orbit_angles.csv');
}
